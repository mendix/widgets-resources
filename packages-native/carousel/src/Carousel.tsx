import { createElement, Fragment, ReactElement, useCallback, useEffect, useState } from "react";
import { ActivityIndicator, LayoutChangeEvent, Text, View } from "react-native";
import { CarouselProps } from "../typings/CarouselProps";
import { CarouselStyle, defaultCarouselStyle, LayoutStyle } from "./ui/styles";
import { default as NativeCarousel, Pagination } from "react-native-snap-carousel";
import deepmerge from "deepmerge";
import { ObjectItem, ValueStatus } from "mendix";

export const Carousel = (props: CarouselProps<CarouselStyle>): ReactElement => {
    const [sliderDimensions, setSliderDimensions] = useState({
        slider: { width: 0, height: 0 },
        slide: { width: 0, height: 0 }
    });

    const customStyles = props.style ? props.style.filter(o => o != null) : [];

    const styles = deepmerge.all<CarouselStyle>([defaultCarouselStyle, ...customStyles]);

    const layoutSpecificStyle: LayoutStyle =
        props.layout === "fullWidth" ? styles.fullWidthLayout! : styles.cardLayout!;

    const [carouselRef, setCarouselRef] = useState(undefined);

    const [activeSlide, setActiveSlide] = useState(0);

    const [loading, setLoading] = useState(true);

    useEffect(() => {
        if (props.contentSource?.status === ValueStatus.Available) {
            setLoading(false);
        }
    }, [props.contentSource]);

    const onSnap = useCallback((index: number) => {
        setActiveSlide(index);
    }, []);

    const renderItem = useCallback(({ item, index }: { item: ObjectItem; index: number }) => {
        const viewStyle = layoutSpecificStyle.slideItem;
        if (viewStyle) {
            // We don't want to pass the already processed height to the item container
            delete viewStyle.width;
        }

        return (
            <View key={index} style={{ ...viewStyle }} testID={`${props.name}$content$${index}`} accessible>
                {props.content(item)}
            </View>
        );
    }, []);

    const renderPagination = useCallback(() => {
        if (!props.showPagination || carouselRef === undefined) {
            return null;
        }

        const contentLength = props.contentSource.items!.length;
        const paginationOverflow = contentLength > 5;
        const { pagination } = layoutSpecificStyle;

        const a11yProps = { accessibilityLabel: `${props.name}$pagination` };

        if (paginationOverflow) {
            return (
                <View style={pagination.container} testID={`${props.name}$pagination`}>
                    <Text style={pagination.text}>
                        {activeSlide + 1}/{contentLength}
                    </Text>
                </View>
            );
        }

        return (
            <Pagination
                dotsLength={contentLength}
                activeDotIndex={activeSlide}
                containerStyle={pagination.container}
                dotContainerStyle={pagination.dotContainerStyle}
                dotColor={pagination.dotStyle?.color}
                dotStyle={pagination.dotStyle}
                inactiveDotStyle={pagination.inactiveDotStyle}
                inactiveDotColor={pagination.inactiveDotStyle?.color}
                inactiveDotOpacity={pagination.inactiveDotStyle?.opacity}
                inactiveDotScale={pagination.inactiveDotStyle?.scale}
                carouselRef={carouselRef}
                tappableDots
                {...a11yProps}
            />
        );
    }, [activeSlide, carouselRef, props.contentSource, props.showPagination]);

    const onLayout = (event: LayoutChangeEvent) => {
        let viewHeight = event.nativeEvent.layout.height;
        const viewWidth = event.nativeEvent.layout.width;

        let itemWidth = 0;
        let itemHeight = 0;

        if (layoutSpecificStyle.slideItem) {
            const { width: slideItemWidth, height: slideItemHeight } = layoutSpecificStyle.slideItem;
            // We calculate the actual number value in order to
            // allow users to set width and height as percentage since lib only accepts numbers

            if (typeof slideItemWidth === "string" && slideItemWidth.includes("%")) {
                const percentage = +slideItemWidth.replace("%", "");
                itemWidth = (viewWidth * percentage) / 100;
            } else {
                itemWidth = Number(slideItemWidth);
            }
            if (typeof slideItemHeight === "string" && slideItemHeight.includes("%")) {
                const percentage = +slideItemHeight.replace("%", "");
                itemHeight = (viewWidth * percentage) / 100;
            } else {
                itemHeight = Number(slideItemHeight);
            }

            if (styles.container?.height === undefined && itemHeight > 0) {
                viewHeight = itemHeight;
            }
        }

        setSliderDimensions({
            slider: { width: viewWidth, height: viewHeight },
            slide: { width: itemWidth, height: itemHeight }
        });
    };

    return (
        <View style={styles.container} onLayout={onLayout}>
            {loading ? (
                <ActivityIndicator color={layoutSpecificStyle.indicator!.color} size="large" />
            ) : (
                /* library ignores width/height if its vertical/horizontal */
                sliderDimensions.slide.width > 0 &&
                sliderDimensions.slider.width > 0 &&
                props.contentSource &&
                props.contentSource.items &&
                props.contentSource.items?.length > 0 && (
                    <Fragment>
                        <NativeCarousel
                            testID={props.name}
                            activeSlideAlignment={props.activeSlideAlignment}
                            layout="default"
                            firstItem={0}
                            useScrollView
                            enableSnap
                            data={props.contentSource.items}
                            renderItem={renderItem}
                            sliderWidth={sliderDimensions.slider.width}
                            sliderHeight={sliderDimensions.slider.height}
                            itemWidth={sliderDimensions.slide.width}
                            itemHeight={sliderDimensions.slide.height}
                            inactiveSlideScale={layoutSpecificStyle.inactiveSlideItem?.scale}
                            inactiveSlideOpacity={layoutSpecificStyle.inactiveSlideItem?.opacity}
                            onSnapToItem={onSnap}
                            ref={(r: any) => setCarouselRef(r)}
                        />
                        {renderPagination()}
                    </Fragment>
                )
            )}
        </View>
    );
};
