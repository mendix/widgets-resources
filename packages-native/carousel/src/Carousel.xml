<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.carousel.Carousel" pluginWidget="true" needsEntityContext="true" offlineCapable="true"
        supportedPlatform="Native"
        xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../node_modules/mendix/custom_widget.xsd">
    <name>Carousel</name>
    <description/>
    <icon>iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAYAAAHdbkFIAAAABGdBTUEAALGPC/xhBQAAADhlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAAqACAAQAAAABAAAAQKADAAQAAAABAAAAQAAAAABlmWCKAAAB9UlEQVR4Ae2aS07DMBCG46iLZMMJYMldmopHWXMLLgBcgFuwpkDV9C4s4QRskkWVkL9oKifNO7jU7W+psj22Z8ZfJolj13F2k8Yv0QMsjWdRKr91PZO7KNSlxg51g/emTQkDRzn3Oa9S53F54x8Lh9zUtysKog2qrAw0CJntrr+S5dTfoG2FsUqRyAeH22AF4glzqwnkQrlLGGPWBxLKg++Fv1eQXZavurhyXfVd1842EjhGAuvHGSauv51Rxw2zuPaeUA7e4iBN0gXKbRNuyHDqn6H/xXt8vlqlt/pYPApRr3wUJEl6IgO6Gse4bG1xKuOLxkWOvNIBvZPJMh0gAZPxRd0kQAIkQAIkYAeBykXpaKSe55feB6YRzKJPfY3XZmrKVZPwygvRd/Ia3+lrTMgaF6X6QrKrcRjQF7JF42iXxBURCZAACZDA/hIYuqXYtGUp7wLmJEACJEACJEACJEACJEACJEAC/0Vgsz8pDhQPz0Wu5/qBusj77GHK2K65fiAvY8v2QKVNz2VfVGS9PsrL9jv77KGKE13zMltlPrXR2wtAG8W29CEAW66UKT8ZAabI2qKXEWDLlTLlJyPAFFlb9DICbLlSpvzsHAH4EMIfOYoO4c8ZuzgTgw3YKtqHT0PP9Io6WScBEiABEiABEiCBgybwA06NjAe1lcwrAAAAAElFTkSuQmCC</icon>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Data source">
                <property key="contentSource" type="datasource" isList="true">
                    <caption>Source</caption>
                    <description/>
                </property>
                <property key="content" type="widgets" dataSource="contentSource">
                    <caption>Content</caption>
                    <description/>
                </property>
            </propertyGroup>
            <propertyGroup caption="Display">
                <property key="layout" type="enumeration" defaultValue="card">
                    <caption>Layout</caption>
                    <description>Define the way items are rendered and animated.</description>
                    <enumerationValues>
                        <enumerationValue key="card">Card</enumerationValue>
                        <enumerationValue key="fullWidth">Full width</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="showPagination" type="boolean" defaultValue="true">
                    <caption>Pagination</caption>
                    <description>Show pagination indicator; dotted for 5 or less items, else show the number of items</description>
                </property>
                <property key="activeSlideAlignment" type="enumeration" defaultValue="center">
                    <caption>Card alignment</caption>
                    <description>Align current card to provide space to display next and previous card. (Only available for Layout type card)</description>
                    <enumerationValues>
                        <enumerationValue key="center">Center</enumerationValue>
                        <enumerationValue key="start">Beginning</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
<!--            Library has a bug with loops-->
<!--            https://github.com/archriss/react-native-snap-carousel/issues/653 - -->
<!--            https://github.com/archriss/react-native-snap-carousel/issues/608-->
<!--            Also not working properly with Mobx-->
<!--            <propertyGroup caption="Behavior">-->
<!--                <property key="loop" type="boolean" defaultValue="false">-->
<!--                    <caption>Loop</caption>-->
<!--                    <description>After the last item the carousel continues to the first item.</description>-->
<!--                </property>-->
<!--            </propertyGroup>-->
            <propertyGroup caption="Common">
                <systemProperty key="Name"/>
                <systemProperty key="Visibility"/>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
