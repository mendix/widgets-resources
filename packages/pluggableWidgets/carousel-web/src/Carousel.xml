<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.custom.carousel.Carousel" pluginWidget="true" needsEntityContext="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd">
    <name>Carousel</name>
    <description>Carousel displaying images</description>
    <studioProCategory>Display</studioProCategory>
    <studioCategory>Display</studioCategory>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Data Source">
                <property key="dataSource" type="datasource" isList="true" required="false">
                    <caption>Data source</caption>
                    <description />
                </property>
                <property key="content" type="widgets" dataSource="dataSource" required="false">
                    <caption>Content</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Display">
                <property key="showPagination" type="boolean" defaultValue="true">
                    <caption>Pagination</caption>
                    <description />
                </property>
                <property key="navigation" type="boolean" defaultValue="true">
                    <caption>Navigation controls</caption>
                    <description />
                </property>
                <property key="autoplay" type="boolean" defaultValue="true">
                    <caption>Auto play</caption>
                    <description />
                </property>
                <property key="delay" type="integer" defaultValue="1000">
                    <caption>Delay</caption>
                    <description>The amount of time to delay between automatically cycling an item (ms)</description>
                </property>
                <property key="loop" type="boolean" defaultValue="true">
                    <caption>Infinite loop</caption>
                    <description />
                </property>
                <property key="animation" type="boolean" defaultValue="true">
                    <caption>Animation</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Events">
                <property key="onClickAction" type="action" required="false">
                    <caption>On click action</caption>
                    <description />
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
