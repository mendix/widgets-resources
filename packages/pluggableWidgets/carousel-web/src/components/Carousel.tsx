import { createElement, ReactNode, ReactElement } from "react";
import { GUID } from "mendix";
import classNames from "classnames";
import { SwiperOptions } from "swiper";
import { Swiper as SwiperReact, SwiperSlide } from "swiper/react";

export interface CarouselProps {
    pagination: boolean;
    loop: boolean;
    animation?: boolean;
    autoplay?: boolean;
    delay?: number;
    navigation: boolean;
    className: string;
    tabIndex?: number | undefined;
    onClick?: () => void;
    items: Array<{ id: GUID; content?: ReactNode }>;
}

export function Carousel(props: CarouselProps): ReactElement {
    const { items, pagination, loop, animation, autoplay, delay, navigation, className, tabIndex, onClick } = props;

    const options: SwiperOptions = {
        slidesPerView: 1,
        centeredSlides: true,
        loop,
        navigation,
        autoplay: autoplay ? { delay, stopOnLastSlide: true } : false,
        pagination: pagination ? { type: "bullets", clickable: true } : false,
        ...(animation && {
            effect: "fade",
            fadeEffect: { crossFade: true }
        })
    };
    return (
        <div className={classNames(className, "widget-carousel")} tabIndex={tabIndex}>
            <SwiperReact {...options} onClick={onClick}>
                {items?.map(item => (
                    <SwiperSlide key={item.id}>{item.content}</SwiperSlide>
                ))}
            </SwiperReact>
        </div>
    );
}
