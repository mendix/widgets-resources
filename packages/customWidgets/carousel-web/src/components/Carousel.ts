import { Component, MouseEventHandler, SFCElement, createElement, ReactNode } from "react";
import classNames from "classnames";

import { Alert } from "./Alert";
import { CarouselControl, CarouselControlProps } from "./CarouselControl";
import { CarouselItem, CarouselItemProps, ItemStatus } from "./CarouselItem";

import "../ui/Carousel.scss";

interface Image {
    guid?: string;
    url: string;
    onClickMicroflow?: string;
    onClickNanoflow?: Nanoflow;
    onClickForm?: string;
    openPageAs: PageLocation;
}

export interface Nanoflow {
    nanoflow: object[];
    paramsSpec: { Progress: string };
}

interface CarouselProps {
    alertMessage?: string;
    images: Image[];
    onClickAction?: (image: Image) => void;
    className?: string;
    style?: object;
}

interface CarouselState {
    activeIndex: number;
    alertMessage?: string;
    showControls?: boolean;
    position?: number;
    animate?: boolean;
}

interface CustomEvent extends Event {
    detail: {
        originPageX: number;
        originPageY: number;
        pageX: number;
        pageY: number;
    };
    target: HTMLElement;
}

type Direction = "right" | "left";
export type PageLocation = "content" | "popup" | "modal";

class Carousel extends Component<CarouselProps, CarouselState> {
    static defaultProps: CarouselProps = {
        images: []
    };
    private moveToTheLeft: MouseEventHandler<HTMLDivElement>;
    private moveToTheRight: MouseEventHandler<HTMLDivElement>;
    private handleSwipeRightEnd: (event: Event) => void;
    private handleSwipeLeftEnd: (event: Event) => void;
    private carouselWidth: number;
    private carouselItems: HTMLElement[];
    private onClickAction: (image: Image) => () => void;
    private swipeStartPosition: number | undefined;

    constructor(props: CarouselProps) {
        super(props);

        this.carouselWidth = 0;
        this.carouselItems = [];
        this.state = {
            activeIndex: 0,
            alertMessage: props.alertMessage,
            animate: false,
            position: 0,
            showControls: this.props.images.length > 1
        };

        this.moveToTheLeft = () => this.moveInDirection("left");
        this.moveToTheRight = () => this.moveInDirection("right");
        this.handleSwipe = this.handleSwipe.bind(this);
        this.addCarouselItem = this.addCarouselItem.bind(this);
        this.handleSwipeLeftEnd = (event: CustomEvent) => this.handleSwipeEnd(event, "left");
        this.handleSwipeRightEnd = (event: CustomEvent) => this.handleSwipeEnd(event, "right");
        this.onClickAction = image => () => {
            if (props.onClickAction) {
                props.onClickAction(image);
            }
        };
    }

    render(): ReactNode {
        const { className, images, style } = this.props;

        return createElement(
            "div",
            { className: classNames("widget-carousel-wrapper", className), style },
            createElement(Alert, {
                bootstrapStyle: "danger",
                className: "widget-carousel-alert",
                message: this.state.alertMessage
            }),
            createElement(
                "div",
                { className: "widget-carousel" },
                createElement(
                    "div",
                    {
                        className: classNames("widget-carousel-item-wrapper", { animate: this.state.animate }),
                        style: { transform: `translate3d(${this.state.position}%, 0px, 0px)` }
                    },
                    this.createCarouselItems(images, this.state.activeIndex || 0)
                ),
                this.createCarouselControls()
            )
        );
    }

    UNSAFE_componentWillReceiveProps(newProps: CarouselProps): void {
        if (this.carouselItems.length) {
            this.removeEvents();
            this.carouselItems = [];
        }

        this.setState({ alertMessage: newProps.alertMessage, showControls: newProps.images.length > 1 });
    }

    componentWillUnmount(): void {
        this.removeEvents();
    }

    private createCarouselItems(images: Image[], activeIndex: number): Array<SFCElement<CarouselItemProps>> {
        return images.map((image, index) => {
            const { position, status } = this.getItemStatus(index, activeIndex);
            return createElement(CarouselItem, {
                getItemNode: this.addCarouselItem,
                key: index,
                onClick: this.onClickAction(image),
                position,
                status,
                url: image.url
            });
        });
    }

    private getItemStatus(index: number, activeIndex: number): { position: number; status: ItemStatus } {
        const maxPercentage = 100;
        return {
            position: (index - activeIndex) * maxPercentage,
            status: classNames({
                active: index === activeIndex,
                next: activeIndex < index,
                prev: activeIndex > index
            }) as ItemStatus
        };
    }

    private createCarouselControls(): Array<SFCElement<CarouselControlProps>> | null {
        if (!this.state.showControls) {
            return null;
        }

        const directions: Direction[] =
            this.state.activeIndex === this.props.images.length - 1
                ? ["left"]
                : this.state.activeIndex === 0
                ? ["right"]
                : ["right", "left"];

        return directions.map((direction, index) =>
            createElement(CarouselControl, {
                direction,
                key: index,
                onClick: direction === "right" ? this.moveToTheRight : this.moveToTheLeft
            })
        );
    }

    private addCarouselItem(carouselItem: HTMLElement | null): void {
        if (carouselItem && this.carouselItems.length < this.props.images.length) {
            this.carouselItems.push(carouselItem);
            this.registerEvents(carouselItem);
        }
    }

    private moveInDirection(direction: Direction, swiping = false): void {
        const { activeIndex } = this.state;
        const newActiveIndex = direction === "right" ? activeIndex + 1 : activeIndex - 1;

        this.setState({
            activeIndex: newActiveIndex,
            alertMessage: "",
            animate: true,
            // eslint-disable-next-line react/no-access-state-in-setstate
            position: swiping ? this.state.position : 0
        });
    }

    private registerEvents(carouselItem: HTMLElement): void {
        carouselItem.addEventListener("swipeleft", this.handleSwipe);
        carouselItem.addEventListener("swipeleftend", this.handleSwipeLeftEnd);
        carouselItem.addEventListener("swiperight", this.handleSwipe);
        carouselItem.addEventListener("swiperightend", this.handleSwipeRightEnd);
        carouselItem.addEventListener("touchmove", event => event.preventDefault());
    }

    private removeEvents(): void {
        this.carouselItems.forEach((carouselItem: HTMLElement) => {
            carouselItem.removeEventListener("swipeleft", this.handleSwipe);
            carouselItem.removeEventListener("swipeleftend", this.handleSwipeLeftEnd);
            carouselItem.removeEventListener("swiperight", this.handleSwipe);
            carouselItem.removeEventListener("swiperightend", this.handleSwipeRightEnd);
            carouselItem.removeEventListener("touchmove", event => event.preventDefault());
        });
    }

    private handleSwipe(event: CustomEvent): void {
        const { parentElement } = event.target;
        this.carouselWidth = this.carouselWidth || (parentElement ? parentElement.offsetWidth : 0);
        const currentPercentage = this.calculateSwipePercentage(event, this.carouselWidth);
        if (!this.shouldSwipe(currentPercentage)) {
            return;
        }
        this.setState({
            // eslint-disable-next-line react/no-access-state-in-setstate
            activeIndex: this.state.activeIndex,
            animate: false,
            position: currentPercentage,
            showControls: false
        });
    }

    private handleSwipeEnd(event: CustomEvent, direction: Direction): void {
        const swipeOutThreshold = 20;
        const currentPercentage = this.calculateSwipePercentage(event, this.carouselWidth);
        if (this.shouldSwipe(currentPercentage)) {
            this.carouselWidth = 0;
            const swipingOut = Math.abs(currentPercentage) > swipeOutThreshold;
            if (swipingOut) {
                this.moveInDirection(direction === "right" ? "left" : "right", true);
            }
        }

        this.swipeStartPosition = 0;
        this.setState({
            // eslint-disable-next-line react/no-access-state-in-setstate
            activeIndex: this.state.activeIndex,
            animate: true,
            position: 0,
            showControls: !!this.carouselItems.length
        });
    }

    private calculateSwipePercentage(event: CustomEvent, width: number): number {
        const maxPercentage = 100;
        if (!this.swipeStartPosition) {
            this.swipeStartPosition = event.detail.pageX;
        }
        const swipeOffset = event.detail.pageX - this.swipeStartPosition;

        return (maxPercentage / width) * swipeOffset;
    }

    private shouldSwipe(percentage: number): boolean {
        return percentage > 0 ? this.state.activeIndex > 0 : this.state.activeIndex < this.carouselItems.length - 1;
    }
}

export { Carousel, CarouselProps, Image };
