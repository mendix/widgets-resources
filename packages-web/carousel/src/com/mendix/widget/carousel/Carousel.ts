import * as dojoDeclare from "dojo/_base/declare";
import * as WidgetBase from "mxui/widget/_WidgetBase";

import { createElement } from "react";
import { render, unmountComponentAtNode } from "react-dom";

import { Carousel, Image } from "./components/Carousel";

class CarouselDojo extends WidgetBase {
    // Properties from Mendix modeler
    staticImages?: Image[];

    postCreate() {
        render(createElement(Carousel, { images: this.staticImages }), this.domNode);
    }

    uninitialize(): boolean {
        unmountComponentAtNode(this.domNode);

        return true;
    }
}

// Declare widget prototype the Dojo way
// Thanks to https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/dojo/README.md
// tslint:disable : only-arrow-functions
dojoDeclare("com.mendix.widget.carousel.Carousel", [ WidgetBase ], function(Source: any) {
    let result: any = {};
    for (let property in Source.prototype) {
        if (property !== "constructor" && Source.prototype.hasOwnProperty(property)) {
            result[property] = Source.prototype[property];
        }
    }
    return result;
}(CarouselDojo));
