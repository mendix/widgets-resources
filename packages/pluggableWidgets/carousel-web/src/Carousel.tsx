import { createElement, ReactNode, useCallback, ReactElement, useEffect } from "react";
import { ValueStatus, GUID, ObjectItem } from "mendix";
import { executeAction } from "@mendix/piw-utils-internal";
import { CarouselContainerProps } from "../typings/CarouselProps";
import { Carousel as CarouselComponent } from "./components/Carousel";
import loadingCircleSvg from "./ui/loading-circle.svg";
import classNames from "classnames";
import "./ui/Carousel.scss";
import Swiper, { Navigation, Pagination, EffectFade, Autoplay } from "swiper";

export function Carousel(props: CarouselContainerProps): ReactNode {
    const { showPagination, loop, tabIndex, navigation, animation, delay, autoplay } = props;
    const onClick = useCallback(() => executeAction(props.onClickAction), [props.onClickAction]);

    useEffect(() => Swiper.use([Navigation, Pagination, EffectFade, Autoplay]), []);

    const renderCarousel = (): ReactElement => {
        return (
            <CarouselComponent
                className={props.class}
                tabIndex={tabIndex}
                pagination={showPagination}
                loop={loop}
                animation={animation}
                autoplay={autoplay}
                delay={delay}
                navigation={navigation}
                items={
                    props.dataSource?.items?.map((item: ObjectItem) => ({
                        id: item.id as GUID,
                        content: props.content?.get(item)
                    })) ?? []
                }
                onClick={onClick}
            />
        );
    };
    const renderLoading = (): ReactNode => {
        return (
            <div className={classNames(props.class, "widget-carousel")} tabIndex={tabIndex}>
                <img src={loadingCircleSvg} className="widget-carousel-loading-spinner" alt="" aria-hidden />
            </div>
        );
    };

    return props.dataSource?.status !== ValueStatus.Available ? renderLoading() : renderCarousel();
}
