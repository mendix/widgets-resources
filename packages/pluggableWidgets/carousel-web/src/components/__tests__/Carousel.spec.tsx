import { createElement } from "react";
import { Carousel, CarouselProps } from "../Carousel";
import renderer from "react-test-renderer";
import { GUID } from "mendix";

describe("Carousel", () => {
    const defaultCarouselProps: CarouselProps = {
        className: "",
        items: [{ id: "1" as GUID, content: <div /> }],
        pagination: true,
        animation: true,
        autoplay: true,
        delay: 3000,
        loop: true,
        navigation: true,
        onClick: () => jest.fn()
    };

    it("renders correctly", () => {
        const carousel = renderer.create(<Carousel {...defaultCarouselProps} />);

        expect(carousel).toMatchSnapshot();
    });
    it("renders correctly without pagination", () => {
        const carousel = renderer.create(<Carousel {...defaultCarouselProps} pagination={false} />);

        expect(carousel).toMatchSnapshot();
    });
    it("renders correctly without navigation", () => {
        const carousel = renderer.create(<Carousel {...defaultCarouselProps} navigation={false} />);

        expect(carousel).toMatchSnapshot();
    });
});
