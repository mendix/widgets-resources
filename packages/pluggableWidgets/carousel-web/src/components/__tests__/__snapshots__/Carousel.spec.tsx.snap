// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel renders correctly 1`] = `
<div
  className="widget-carousel"
>
  <div
    className="swiper-container"
  >
    <div
      className="swiper-button-prev"
    />
    <div
      className="swiper-button-next"
    />
    <div
      className="swiper-pagination"
    />
    <div
      className="swiper-wrapper"
    >
      <div
        className="swiper-slide swiper-slide-duplicate"
        data-swiper-slide-index={0}
      >
        <div />
      </div>
      <div
        className="swiper-slide"
        data-swiper-slide-index={0}
      >
        <div />
      </div>
      <div
        className="swiper-slide swiper-slide-duplicate"
        data-swiper-slide-index={0}
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`Carousel renders correctly without navigation 1`] = `
<div
  className="widget-carousel"
>
  <div
    className="swiper-container"
  >
    <div
      className="swiper-pagination"
    />
    <div
      className="swiper-wrapper"
    >
      <div
        className="swiper-slide swiper-slide-duplicate"
        data-swiper-slide-index={0}
      >
        <div />
      </div>
      <div
        className="swiper-slide"
        data-swiper-slide-index={0}
      >
        <div />
      </div>
      <div
        className="swiper-slide swiper-slide-duplicate"
        data-swiper-slide-index={0}
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`Carousel renders correctly without pagination 1`] = `
<div
  className="widget-carousel"
>
  <div
    className="swiper-container"
  >
    <div
      className="swiper-button-prev"
    />
    <div
      className="swiper-button-next"
    />
    <div
      className="swiper-wrapper"
    >
      <div
        className="swiper-slide swiper-slide-duplicate"
        data-swiper-slide-index={0}
      >
        <div />
      </div>
      <div
        className="swiper-slide"
        data-swiper-slide-index={0}
      >
        <div />
      </div>
      <div
        className="swiper-slide swiper-slide-duplicate"
        data-swiper-slide-index={0}
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;
